src/include/parser/create_function_statement.h

```diff
@@ -100,8 +85,8 @@ class CreateFunctionStatement : public SQLStatement {
    * @param as_type executable or query string
    */
   CreateFunctionStatement(bool replace, std::string func_name, std::vector<std::string> func_body,
-                          std::shared_ptr<ReturnType> return_type,
-                          std::vector<std::shared_ptr<FuncParameter>> func_parameters, PLType pl_type, AsType as_type)
+                          std::unique_ptr<ReturnType> return_type,
+                          std::vector<std::unique_ptr<FuncParameter>> func_parameters, PLType pl_type, AsType as_type)
```

Comment:
Why is this required to be a unique ptr? This structure isn't storing the pointer anyway and it's just moving the data pointed by it. Is it a way of enforcing that func_parameters is deallocated by who ever is passing it in?

Reply:
Unique pointers seem like a good idea to me, it's enforcing the deallocation in code. I was just clarifying if that was the intended goal.

Reply:
If I understand correctly, your question @yash620 is "why is the class storing a pointer to the thing instead of storing the thing directly".

This is also a question that I had back when I originally ported the parser last year. I believe a meta reason for keeping it this way is that the rest of the Peloton system (optimizer, binder, etc) would have expected it to remain this way, and this minimized friction for future porting.

If the user needs another reference to a unique pointer, they should be able to get a managed pointer from it. Let me know if you see somewhere like that. We do want to avoid accidentally doing costly deep copies everywhere, supposedly that's how Peloton became a lot slower.

But to return to the original question; I think with the introduction of managed pointers there is actually an advantage to keeping it as a unique pointer. Now, when someone asks you for your ReturnType or whatever, you hand off a managed pointer and it remains a managed pointer and you know that you don't own it.

It is possible that your question is instead "why isn't it a vector<FuncParameter>". I _think_ the optimizer and binder would prefer it to be vector<unique_ptr<FuncParameter>>, but I could be mistaken.

Reply:
Long story short; the parser was initially ported without any customers, so we kept a lot of internal representation the same to make it easier to port the customers over. That said, if a refactor makes sense, we should probably do it.